### ISSUE TITLE
Appropriate license for external C files

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1525]
**Reported by beutlich on 24 Jun 2014 20:09 UTC**
This one came up when [evaluating the Noise lib](https://trac.modelica.org/Modelica/ticket/1514) for MSL potential integration. See also https://github.com/DLR-SR/Noise/issues/9 and https://github.com/modelica/Modelica_DeviceDrivers/issues/35.

Is it right to have the external C source and header files licensed under Modelica License 2? Usually only the Modelica files of a package use Modelica License 2 whereas C source and header files use some more appropriate license like BSD3.

Current state: In MSL there are ModelicaInternal.c and ModelicaStrings.c licensed under Modelica License 2 whereas ModelicaStandardTables.\* are ModelicaMatIO.\* are licensed under BSD3. win32_dirent.\* is public domain.

In the MSL readme.txt it treads
```
All files in this directory (Modelica) and in all subdirectories,
especially in Modelica/Resources/* are licensed
by the Modelica Association under the Modelica License 2
```
thus with emphasis on the C sources. What is the reason that Modelica License 2 is used here? Or is it just a mistake in the DLR library template?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1525

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 25 Jun 2014 05:03 UTC**
Replying to [ticket:1525 beutlich]:
> In the MSL readme.txt it treads
> ```
> All files in this directory (Modelica) and in all subdirectories,
> especially in Modelica/Resources/\* are licensed
> by the Modelica Association under the Modelica License 2
> ```

(It also lists the tables C-Sources as exceptions. Same thing in the package.mo, which contains the license).

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 25 Jun 2014 07:26 UTC**
Replying to [comment:1 sjoelund.se]:
> Replying to [ticket:1525 beutlich]:
> > In the MSL readme.txt it treads
> > {{{
> > All files in this directory (Modelica) and in all subdirectories,
> > especially in Modelica/Resources/\* are licensed
> > by the Modelica Association under the Modelica License 2
> > }}}
> 
> (It also lists the tables C-Sources as exceptions. Same thing in the package.mo, which contains the license).

Yes I know. Forgot to mention it here since focus is on ModelicaInternal.c and ModelicaStrings.c.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 5 Aug 2014 10:29 UTC**
Any comments from map-lib deciders?

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 19 Aug 2014 08:25 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 19 Dec 2014 07:01 UTC**
Ticket retargeted after milestone closed

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 20 Feb 2015 14:42 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Feb 2015 17:34 UTC**
Modelica design meeting:
No conclusion.

The problem is you want to take C-code from MSL and use it in an unrelated project.
Using BSD (3?) seems like a solution.

Also problem with maintaining unmaintained Modelica libraries - since we want to upgrade (not a bug fix) and change version but not the name.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 15 Apr 2015 14:33 UTC**
Ticket retargeted after milestone closed

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 5 May 2015 13:24 UTC**
Replying to [comment:7 hansolsson]:
> The problem is you want to take C-code from MSL and use it in an unrelated project.
> Using BSD (3?) seems like a solution.

Yes, changing the license of the external C-code of Modelica libraries like Modelica_DeviceDrivers or Noise to BSD3 looks correct to me.
> 
> Also problem with maintaining unmaintained Modelica libraries - since we want to upgrade (not a bug fix) and change version but not the name.

The same naming issues already exists, if the (Git) repository of a Modelica library under Modelica License 2 is forked.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 26 Jun 2015 14:54 UTC**
Ticket retargeted after milestone closed

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 6 Aug 2015 07:27 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 24 Sep 2015 10:28 UTC**
Discussion during 87th meeting in Paris:
Change license from Modelica License 2 to BSD 3-clause 
Poll:	In favour:	12
 	Against:	1
 	Abstain:	10
Martin formulates a proposal and explains why, 
Toni starts electronic voting within MAP-Lib

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 24 Sep 2015 21:53 UTC**
There currently is an issue with the license of the FMI 2.0 files, see [#351](https://trac.fmi-standard.org/ticket/351). As pointed out at the 87. MDM the license of the FMI files and the external C files of the MSL should be available under the same license. Thus it might be also be necessary to change the license to BSD 2-Clause here for files ModelicaStandardTables.[ch], ModelicaStrings.c and ModelicaInternal.c.

Edit: ModelicaRandom.c (of Modelica Noise) currently is licensed under BSD 2-Clause.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 Sep 2015 20:33 UTC**
Replying to [comment:12 ahaumer]:
> Martin formulates a proposal and explains why, 
> Toni starts electronic voting within MAP-Lib

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 6 Oct 2015 08:54 UTC**
Martin, when do you plan to formulate the proposal? As stated in comment:13 the used licenses are currently pretty messy. Actually I'd like to see BSD 2-Clause but at the 87. MDM we voted on BSD 3-Clause.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 7 Oct 2015 07:48 UTC**
2-clause or 3-clause makes only a small difference. But considering we want Modelica to be spread, it would make sense to not gag people from saying the MA developed ModelicaStandardTables, etc.

The added clause:
> Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 7 Oct 2015 07:58 UTC**
I think there is pretty much agreement here to not have the third BSD clause.

 * FMI files are BSD 2-Clause (they should be)
 * ModelicaRandom.c is BSD 2-Clause
 * ModelicaMatIO.[ch] is BSD 2-Clause

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 7 Oct 2015 10:39 UTC**
Replying to [comment:16 sjoelund.se]:
> 2-clause or 3-clause makes only a small difference. But considering we want Modelica to be spread, it would make sense to not gag people from saying the MA developed ModelicaStandardTables, etc.

I don't see the point is about 'MA developed ModelicaStandardTables'; but about what you say about your own product using ModelicaStandardTables - and its contributors.

That is I thought the goal of 3rd clause was more that people are not allowed to say 
'MyModelicaTool - using the latest ModelicaStandardTables developed by ITI'
'MyModelicaTool - the official Modelica Association tool',

I can see that these can be problematic, and I am not sure you are allowed do that even without the 3rd clause (unless you have permission). [Not saying that they make sense from a marketing perspective.] On the other hand a tool could just mention that 'MA developed ModelicaStandardTables' in some non-endorsing way; which should be ok.

The endorsement part and name of contributors seems the most problematic; note that some Creative Commons Licenses also include non-endorsement texts, as well as clause 7 of Modelica License 2. The Creative Commons is interesting since CC-BY-3.0 has a non-endorsement clause, but CC-BY-4.0 states that the license does not grant you the endorsement right.

> The added clause:
> > Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 8 Oct 2015 10:10 UTC**
Although guilty myself of voting for the BSD-3 clause back in Velizy , now, after I had actually some time to reflect the implications I think we should be pragmatic and ask ourself _if_ something like this is going to happen:


Replying to [comment:18 hansolsson]:
> That is I thought the goal of 3rd clause was more that people are not allowed to say 
> 'MyModelicaTool - using the latest ModelicaStandardTables developed by ITI'
> 'MyModelicaTool - the official Modelica Association tool',

will we as MA actually put energy and resources into following up this kind of miss-endorsement. And I think the answer is _no_. So we might just as well drop that third clause and be honest with ourselves and go for the BSD-2 clause licence.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 8 Oct 2015 10:20 UTC**
I am also strongly for the BSD 2 clause license. Reasons:

- The third clause of the BSD 3 clause license is difficult to understand and also the consequences are not obvious.

- I have seen presentations about "open source licenses" where the advice was given, to not mention at all the authors of the open source product, in order to not violate the third clause (although in my opinion this is wrong and violates the other clause). This would be completely against the intention of the BSD clause: "do with it what you want, but give credits to me and do not make me responsible if there are bugs".

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 8 Oct 2015 10:36 UTC**
Fine that there is some agreement on using the BSD 2-Clause license. How to continue?

 1. @otter: Are you going to formulate some proposal and we accept it without voting?
 1. @otter: Are you going to formulate some proposal and set up some voting?
 1. Should we ignore the Velizy voting and change the licensing according to our all new understanding?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 8 Oct 2015 10:38 UTC**
Replying to [comment:20 otter]:
> I am also strongly for the BSD 2 clause license. Reasons:
> 
> - The third clause of the BSD 3 clause license is difficult to understand and also the consequences are not obvious.

I agree that the consequences are not obvious - but to me this applies also to BSD 2 clause, and I would hope that some lawyer could resolve this.

To me it would be logical that BSD 2-clause doesn't grant me the right to use contributors etc to endorse and promote a product.

So, why did someone think it necessary to include the 3rd clause? One could think it was just a mistake a long time ago, but Creative Commons also use various non-endorsements clauses.

The Creative Commons in their FAQ mention that they make the non-endorsement 'explicit'; sort of indicating that it is implicit in other cases and it just seems like a way to cover themselves.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 8 Oct 2015 11:12 UTC**
Replying to [comment:22 hansolsson]:
> Replying to [comment:20 otter]:
> > I am also strongly for the BSD 2 clause license. Reasons:
> > 
> > - The third clause of the BSD 3 clause license is difficult to understand and also the consequences are not obvious.
> 
> I agree that the consequences are not obvious - but to me this applies also to BSD 2 clause, and I would hope that some lawyer could resolve this.
> 
> To me it would be logical that BSD 2-clause doesn't grant me the right to use contributors etc to endorse and promote a product.
> 
> So, why did someone think it necessary to include the 3rd clause? One could think it was just a mistake a long time ago, but Creative Commons also use various non-endorsements clauses.
> 
> The Creative Commons in their FAQ mention that they make the non-endorsement 'explicit'; sort of indicating that it is implicit in other cases and it just seems like a way to cover themselves.

For Modelica Association it does not matter, if someone wants to say, "I am using this code from MA (and since it is from MA it is high quality)", or someone just ships the conditions (say, within the MSL). Therefore, whatever interpretation the user has, it does not matter for us. There are so many other high priority issues, that we should not spend time on a really tiny issue where a good solution is available and we just have to copy this into the MSL.

We probably still should have a vote: "BSD 2 clause" yes/no.

Toni, can you organize this. Proposal for text:
  It is proposed to change the license of the C-source code shipped with MSL that has been provided by MA members. The current "Modelica License 2" shall be replaced by the BSD 2-clause license (https://opensource.org/licenses/BSD-2-Clause).
Do you agree: yes/no.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 8 Oct 2015 11:24 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 9 Oct 2015 09:11 UTC**
To me it seems best to first determine the facts, and then take the decision.
After some searching it seems the non-endorsement clause is **basically meaningless**, and there is no practical difference between 3-clause and 2-clause BSD, see e.g.

http://choosealicense.com/licenses/

which seems very reliable and indicates that the non-endorsement is implicit anyway.
(You have to click around a bit to find 3-clause BSD and that text.)

If someone disagrees with this (and prefer 3-clause BSD) they should explain why.

Replying to [comment:19 dietmarw]:
> Although guilty myself of voting for the BSD-3 clause back in Velizy , now, after I had actually some time to reflect the implications I think we should be pragmatic and ask ourself _if_ something like this is going to happen:
> 
> 
> Replying to [comment:18 hansolsson]:
> > That is I thought the goal of 3rd clause was more that people are not allowed to say 
> > 'MyModelicaTool - using the latest ModelicaStandardTables developed by ITI'
> > 'MyModelicaTool - the official Modelica Association tool',
> 
> will we as MA actually put energy and resources into following up this kind of miss-endorsement. And I think the answer is _no_. So we might just as well drop that third clause and be honest with ourselves and go for the BSD-2 clause licence.

Both MA and ITI can in the scenario above still act if their trademarks/logos are used inappropriately (which would be the case above). And there are also other rules against false advertising.

If people just claim that MSL and Modelica are developed by Modelica Association they can use Modelica-logo etc according to our updated rules (independent of this license).

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 27 Oct 2015 13:00 UTC**
> Toni, can you organize this. Proposal for text:
>   It is proposed to change the license of the C-source code shipped with MSL that has been provided by MA members. The current "Modelica License 2" shall be replaced by the BSD 2-clause license (https://opensource.org/licenses/BSD-2-Clause).
> Do you agree: yes/no.
> 

Toni: Bump. There are only about two months left for MSL 3.2.2 release (according to the minutes from the 87. MDM in Velizy) and I do not want to vote on New Year's Eve. So let's please finalize this annoying license ticket if there is even agreement among the panel list. Thanks!

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 19 Nov 2015 10:27 UTC**
The electronic voting gave the following result: 
9 votes, all in favor. 
So you can proceed with the https://opensource.org/licenses/BSD-2-Clause. 
What do we have to do to close the ticket?

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 19 Nov 2015 14:42 UTC**
Replying to [comment:27 ahaumer]:
> The electronic voting gave the following result: 
> 9 votes, all in favor. 

Thanks.
> So you can proceed with the https://opensource.org/licenses/BSD-2-Clause. 
> What do we have to do to close the ticket? 

Change the license comments in the appropriate files and the readme's. Will do it.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 25 Nov 2015 10:48 UTC**
Resolved by ae5252055b23c016787957582c9089ab69805c82.

**Comment by modelica-trac-importer** [comment]
**Changelog modified by beutlich on 25 Nov 2015 10:48 UTC**
Changed license of ModelicaExternalC C sources to BSD 2-Clause License (aka Simplified BSD License)
